sss.go

```diff
@@ -11,19 +11,93 @@ import (
 	"github.com/lestrrat-go/jwx/jwe"
 )
 
-// DecryptSss implements Shamir Secret Sharing decryption algorithm
-func DecryptSss(msg *jwe.Message, clevisNode map[string]interface{}) ([]byte, error) {
-	sssNode, ok := clevisNode["sss"].(map[string]interface{})
-	if !ok {
-		return nil, fmt.Errorf("clevis.go/sss: cannot parse provided token, node 'clevis.sss'")
+// SssPin represents the data samir secret sharing needs to perform decryption
+type SssPin struct {
+	Prime     string   `json:"p"`
+	Threshold int      `json:"t"`
+	Jwe       []string `json:"jwe"`
+}
+
+func appendConfig(slice []json.RawMessage, obj interface{}) ([]json.RawMessage, error) {
+	cfgBytes, err := json.Marshal(obj)
+	if err != nil {
+		return slice, err
 	}
+	return append(slice, cfgBytes), nil
+}
 
-	primeNode, ok := sssNode["p"].(string)
-	if !ok {
-		return nil, fmt.Errorf("clevis.go/sss: cannot parse provided token, node 'clevis.sss.p'")
+// ToConfig converts a given SssPin into the corresponding SssConfig which can be used for encryption
+func (p SssPin) ToConfig() (SssConfig, error) {
+	c := SssConfig{
+		Threshold: p.Threshold,
+		Pins:      make(map[string][]json.RawMessage),
 	}
+	jweErrors := make([]error, 0)
+	for _, jwe := range p.Jwe {
+		_, pin, err := Parse([]byte(jwe))
+		if err != nil {
+			jweErrors = append(jweErrors, err)
+			continue
+		}
+		switch pin.Pin {
+		case "tang":
+			cfg, err := pin.Tang.ToConfig()
+			if err != nil {
+				jweErrors = append(jweErrors, err)
+				continue
+			}
+			c.Pins["tang"], err = appendConfig(c.Pins["tang"], cfg)
+			if err != nil {
+				jweErrors = append(jweErrors, err)
+				continue
+			}
+		case "tpm2":
+			cfg, err := pin.Tpm2.ToConfig()
+			if err != nil {
+				jweErrors = append(jweErrors, err)
+				continue
+			}
+			c.Pins["tpm2"], err = appendConfig(c.Pins["tpm2"], cfg)
+			if err != nil {
+				jweErrors = append(jweErrors, err)
+				continue
+			}
+		case "sss":
+			cfg, err := pin.Sss.ToConfig()
+			if err != nil {
+				jweErrors = append(jweErrors, err)
+				continue
+			}
+			c.Pins["sss"], err = appendConfig(c.Pins["sss"], cfg)
```

Comment:
`appendConfig` repeats in the all `case` branches. For me it looks like it can be moved outside of `switch` statement.

And `appendConfig` can be inlined as it is a trivial function (i.e. just call marshal and append directly without introducing a new function).

Reply:
But since every 'cfg' is a different type, I'm not sure how to do that....  I did have a version of the code that used an interface so I could have common code for any `interface{ ToConfig() (config, error) }`, but the fact that this changes `ToConfig` to have to return an interface as well, which then has other implications elsewhere in the code...

Reply:
I was thinking of making it `var cfg interface{}` and putting it before the `switch`.

Reply:
Great idea!  It's much simpler now.